# Анализ базы данных сервиса чтения книг (SQL)

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Наша первая задача как аналитиков — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Описание данных:**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

**Задачи:**

- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг —  следует учитывать только книги с 50 и более оценками;
- Посчитайть среднее количество обзоров от пользователей, которые поставили больше 50 оценок.


<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подключение-к-базе-данных" data-toc-modified-id="Подключение-к-базе-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подключение к базе данных</a></span></li><li><span><a href="#Загрузка-таблиц" data-toc-modified-id="Загрузка-таблиц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Загрузка таблиц</a></span><ul class="toc-item"><li><span><a href="#Таблица-book" data-toc-modified-id="Таблица-book-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Таблица book</a></span></li><li><span><a href="#Таблица-authors" data-toc-modified-id="Таблица-authors-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Таблица authors</a></span></li><li><span><a href="#Таблица-publishers" data-toc-modified-id="Таблица-publishers-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Таблица publishers</a></span></li><li><span><a href="#Таблица--ratings" data-toc-modified-id="Таблица--ratings-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Таблица  ratings</a></span></li><li><span><a href="#Таблица-reviews" data-toc-modified-id="Таблица-reviews-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Таблица reviews</a></span></li><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку;" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку;-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку;</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры;" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры;-2.8"><span class="toc-item-num">2.8&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками;" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками;-2.9"><span class="toc-item-num">2.9&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок." data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок.-2.10"><span class="toc-item-num">2.10&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

## Подключение к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Загрузка таблиц

###  Таблица book

In [3]:
query = '''
    SELECT * 
    FROM books
'''    
pd.io.sql.read_sql(query, con = engine) 

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

В таблице хранится информация о 1000 книг.

### Таблица authors

In [4]:
query = '''
    SELECT * 
    FROM authors
'''
pd.io.sql.read_sql(query, con = engine) 

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

В таблице информация о 636 авторах.

### Таблица publishers

In [5]:
query = '''
    SELECT * 
    FROM publishers
'''
pd.io.sql.read_sql(query, con = engine) 

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

В таблице информация о 340 издательствах.

### Таблица  ratings

In [6]:
query = '''
    SELECT * 
    FROM ratings
'''
pd.io.sql.read_sql(query, con = engine) 

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

В таблице содержится информация о 6456 пользовательских оценок книг.

### Таблица reviews

In [7]:
query = '''
    SELECT * 
    FROM reviews
'''
pd.io.sql.read_sql(query, con = engine) 

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

В таблице содержится  2793 обзора на книги.

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [8]:
query = '''
SELECT COUNT(DISTINCT book_id) AS count_book 
FROM   books
WHERE  publication_date > '2000-01-01';
'''
pd.io.sql.read_sql(query, con = engine)

count_book
0         819

После 1 января 2000 года вышло 819 книг.

### Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [9]:
query = '''
SELECT books.book_id,
       books.title AS name_book,
       COUNT(DISTINCT reviews.review_id) AS count_review,
       ROUND(AVG(ratings.rating),1) AS avg_rating
FROM   books 
       LEFT JOIN reviews ON reviews.book_id = books.book_id
       LEFT JOIN ratings ON ratings.book_id = books.book_id
GROUP BY books.book_id
ORDER BY count_review DESC;
'''
pd.io.sql.read_sql(query, con = engine)

book_id                                          name_book  count_review  \
0        948                            Twilight (Twilight  #1)             7   
1        963                                Water for Elephants             6   
2        734                                   The Glass Castle             6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...             6   
4        695  The Curious Incident of the Dog in the Night-Time             6   
..       ...                                                ...           ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
996      808                            The Natural Way to Draw             0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites             0   
998      221                          Essential Tales and Poems             0   
999      191  Disney's Beauty and the Beast (A Little Golden...             0   

     avg_rating  
0           3.7  
1           4.0  
2           4.2  
3           4.4  
4           4.1  
..          ...  
995         3.7  
996         3.0  
997         5.0  
998         4.0  
999         4.0  

[1000 rows x 4 columns]

Максимальное число обзорову книги Сумерки - "7", также есть книги, у которых  нет обзоров.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [10]:
query = '''
SELECT publishers.publisher,
       COUNT(book_id) AS books_cnt
FROM   books
       LEFT JOIN publishers ON publishers.publisher_id = books.publisher_id
WHERE num_pages > 50
GROUP BY publishers.publisher
ORDER BY books_cnt DESC
LIMIT 1;
'''
pd.io.sql.read_sql(query, con = engine)

publisher  books_cnt
0  Penguin Books         42

Издательство "Penguin Books" выпустило 42 книги  толщиной более 50 страниц.

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [11]:
query = '''
WITH autor_avg AS (SELECT authors.author AS author, 
                          AVG(ratings.rating) AS avg_rating
                   FROM books
                        LEFT JOIN authors ON books.author_id = authors.author_id
                        LEFT JOIN ratings ON books.book_id = ratings.book_id
                   GROUP BY books.book_id, 
                            authors.author_id
                   HAVING COUNT(ratings.rating_id) >=50 )

SELECT author,
       ROUND(AVG(avg_rating),2) AS avg_autor_rating
FROM  autor_avg 
GROUP BY author
ORDER BY avg_autor_rating DESC
LIMIT 5;
''' 

pd.io.sql.read_sql(query, con = engine)

author  avg_autor_rating
0         J.K. Rowling/Mary GrandPré              4.28
1  Markus Zusak/Cao Xuân Việt Khương              4.26
2                     J.R.R. Tolkien              4.26
3                  Louisa May Alcott              4.19
4                       Rick Riordan              4.08

Автор с самой высокой средней оценкой - Джоан Роулинг. 

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [12]:
query = '''
WITH count_rating AS (SELECT username,
                             COUNT(rating_id) AS ratings_cnt
                       FROM ratings
                       GROUP BY username
                       HAVING COUNT(rating_id) > 50),
     count_reviews AS (SELECT username,
                              COUNT(review_id) AS reviews_cnt
                       FROM reviews
                       GROUP BY username)


SELECT ROUND(AVG(reviews_cnt),2) as avg_reviews
FROM count_rating
     LEFT JOIN count_reviews ON count_reviews.username = count_rating.username
'''
pd.io.sql.read_sql(query, con = engine)

avg_reviews
0        24.33

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24 обзора.

## Вывод
Загрузили все  необходимые таблицы. Ответили на все необходимые вопросы. 